### PR TITLE
Logs attributes

### PR DESCRIPTION
**Pull request by elijahbenizzy**
Logs attributes using the __tracer. Decisions:
1. We can log both with the tracer factory and the tracer. You have to log the right one to get it at the right scope
2. Keys are unique on spanID/action (not enforced yet)
3. Saved to log file as a new entry
4. Visualized in the UI
5. No async implemented yet
6. Uses standard serde (could probably document this better)

See this:
```python
from burr.visibility import TracingFactory
from burr.core import action

@action(reads=['prompt'], writes=['response'])
def my_action(state: State, __tracer: TracingFactory) -> State:
    __tracer.log_attribute(prompt_length=len(state["prompt"]), prompt=state["prompt"])
    with __tracer('create_prompt') as t:
        modified_prompt = _modify_prompt(state["prompt"])
        t.log_attribute(modified_prompt=modified_prompt)
        with __tracer('check_prompt_safety') as t:
            if is_unsafe:=_is_unsafe(modified_prompt):
                modified_prompt = _fix(modified_prompt)
            t.log_attribute(fixed_prompt=modified_prompt, is_unsafe=is_unsafe)
    with __tracer('call_llm', dependencies=['create_prompt']):
        response = _query(prompt=modified_prompt)
        t.log_attribute(response=response.message, tokens=response.tokens)
    return state.update({'response': response.message})
```

Need to get working on s3

## Changes
See above/commit messages.

## How I tested this

Locally + unit tests

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: burr/tracking/client.py [hunk 1]
```diff
@@ -454,6 +458,26 @@ def post_end_span(
...
 
+    async def do_log_attribute(
+        self,
+        *,
+        artifact_name: str,
+        tags: dict,
+        value: Any,
+        action: str,
+        action_sequence_id: int,
+        span: "ActionSpan",
```
Comment (skrawcz) [inline burr/tracking/client.py:469]: where is this used?

**Comment by elijahbenizzy** [comment]
Mid-stride:
OK, todo to get this ready
- [ ] Rebase on #288  -- some step-parsing stuff
- [ ] Add to UI
- [ ] Add tests
- [ ] Add documentation

File: docs/concepts/additional-visibility.rst [hunk 2]
```diff
@@ -17,7 +17,7 @@ the `OpenTelemetry <https://opentelemetry.io/>`_ sdk, although it is simplified
 To add the tracing capability, the action first has to declare a ``__tracer`` input. This is a
 an object that instantiates spans and tracks state.
 
-Then, using the `__tracer` as a callable, you can instantiate spans and track state.
+Then, using the ``__tracer`` as a callable, you can instantiate spans and track state.
```
Comment (skrawcz) [inline docs/concepts/additional-visibility.rst:20]: callable or context manager?

**Review by skrawcz — APPROVED** [review]
looks great.

File: docs/concepts/additional-visibility.rst (hunk 2 above)
Comment (elijahbenizzy) [reply docs/concepts/additional-visibility.rst:20]: Both, actually

**Pull request merged**